File: integ-tests/python/tests/test_client_option.py
import pytest
import baml_py
from ..baml_client import b
from ..baml_client.sync_client import b as sync_b


@pytest.mark.asyncio
async def test_client_option_routes_to_correct_client():
    """Test that the client option routes requests to the specified client."""
    # ExtractResume normally uses GPT4 (openai), but we override to Claude
    request = await b.request.ExtractResume(
        "test resume",
        baml_options={"client": "Claude"}
    )

    # Should route to Anthropic API
    assert "anthropic" in request.url.lower()


@pytest.mark.asyncio
async def test_client_option_takes_precedence_over_client_registry():
    """Test that client option takes precedence over client_registry.set_primary()."""
    cr = baml_py.ClientRegistry()
    cr.set_primary("GPT4")  # This should be overridden

    request = await b.request.ExtractResume(
        "test resume",
        baml_options={"client": "Claude", "client_registry": cr}
    )

    # client option should win - should route to Anthropic, not OpenAI
    assert "anthropic" in request.url.lower()


@pytest.mark.asyncio
async def test_client_registry_still_works():
    """Test that client_registry without client option still works."""
    cr = baml_py.ClientRegistry()
    cr.set_primary("Claude")

    request = await b.request.ExtractResume(
        "test resume",
        baml_options={"client_registry": cr}
    )

    # Should route to Anthropic API
    assert "anthropic" in request.url.lower()


@pytest.mark.asyncio
async def test_with_options_client():
    """Test that with_options(client=...) works correctly."""
    my_b = b.with_options(client="Claude")

    request = await my_b.request.ExtractResume("test resume")

    # Should route to Anthropic API
    assert "anthropic" in request.url.lower()


def test_client_option_sync():
    """Test that the client option works with sync client."""
    request = sync_b.request.ExtractResume(
        "test resume",
        baml_options={"client": "Claude"}
    )

    # Should route to Anthropic API
    assert "anthropic" in request.url.lower()


def test_with_options_client_sync():
    """Test that with_options(client=...) works with sync client."""
    my_b = sync_b.with_options(client="Claude")

    request = my_b.request.ExtractResume("test resume")

    # Should route to Anthropic API
    assert "anthropic" in request.url.lower()

File: integ-tests/rust/tests/test_client_option.rs
//! Client option tests - ported from test_client_option_test.go
//!
//! Tests for client option functionality including:
//! - WithClient overriding
//! - Option precedence
//! - Combination with other options

use baml::ClientRegistry;
use baml::LogType;
use rust::baml_client::new_collector;
use rust::baml_client::sync_client::B;
use std::collections::HashMap;

/// Test with_client routes to correct client - Go: TestWithClientOption/WithClientRoutesToCorrectClient
#[test]
fn test_with_client_override() {
    // Go: Use WithClient to override the default client
    // Claude is defined in the BAML files
    let result = B.TestOpenAI.with_client("Claude").call("Say hello");
    assert!(
        result.is_ok(),
        "Expected successful call with Claude client, got {:?}",
        result
    );
    let output = result.unwrap();
    // Go: assert.NotEmpty(t, result)
    assert!(!output.is_empty(), "Expected non-empty result");
}

/// Test with_client takes precedence over with_client_registry - Go: TestWithClientOption/WithClientTakesPrecedenceOverWithClientRegistry
#[test]
fn test_with_client_precedence_over_registry() {
    // Create a client registry with GPT4oMini as primary
    let mut registry = ClientRegistry::new();
    let mut options = HashMap::new();
    options.insert("model".to_string(), serde_json::json!("gpt-4o-mini"));
    registry.add_llm_client("MyGPT", "openai", options);
    registry.set_primary_client("MyGPT");

    // Go: WithClient should override the client registry's primary
    // Use Claude which is defined in BAML files
    let result = B
        .TestOpenAI
        .with_client_registry(&registry)
        .with_client("Claude")
        .call("Say hello");
    assert!(
        result.is_ok(),
        "Expected successful call with Claude overriding registry, got {:?}",
        result
    );
    let output = result.unwrap();
    // Go: assert.NotEmpty(t, result)
    assert!(!output.is_empty(), "Expected non-empty result");
}

/// Test with_client_registry still works - Go: TestWithClientOption/WithClientRegistryStillWorks
#[test]
fn test_with_client_registry_still_works() {
    let mut registry = ClientRegistry::new();
    let mut options = HashMap::new();
    options.insert("model".to_string(), serde_json::json!("gpt-4o-mini"));
    registry.add_llm_client("MyGPT", "openai", options);
    registry.set_primary_client("MyGPT");

    let result = B
        .TestOpenAI
        .with_client_registry(&registry)
        .call("Say hello");
    assert!(
        result.is_ok(),
        "Expected successful call with registry, got {:?}",
        result
    );
    let output = result.unwrap();
    // Go: assert.NotEmpty(t, result)
    assert!(!output.is_empty(), "Expected non-empty result");
}

/// Test client option with collector - Go: TestWithClientOption/WithClientAndCollector
#[test]
fn test_client_option_with_collector() {
    let collector = new_collector("client-option-test");

    let result = B
        .TestOpenAI
        .with_client("Claude")
        .with_collector(&collector)
        .call("Say hello");

    assert!(
        result.is_ok(),
        "Expected successful call with collector and client, got {:?}",
        result
    );
    let output = result.unwrap();
    // Go: assert.NotEmpty(t, result)
    assert!(!output.is_empty(), "Expected non-empty result");

    // Go: Verify collector captured the call
    let logs = collector.logs();
    // Go: assert.Len(t, logs, 1)
    assert_eq!(logs.len(), 1, "Expected one log entry");

    // Verify the call was logged correctly
    let log = &logs[0];
    assert_eq!(
        log.function_name(),
        "TestOpenAI",
        "Function name should match"
    );
    assert_eq!(log.log_type(), LogType::Call, "Log type should be call");
}
